compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/JvmBuiltinOptimizationLowering.kt

```diff
@@ -43,56 +45,72 @@ class JvmBuiltinOptimizationLowering(val context: JvmBackendContext) : FileLower
-            return (right.isNullConst() && left is IrConst<*>)
-                    || (left.isNullConst() && right is IrConst<*>)
+            return left to right
+        } else if (call.symbol.owner.isObjectEquals) {
+            val left = call.dispatchReceiver!!
+            val right = call.getValueArgument(0)!!
+            return left to right
+        } else {
+            return null;
         }
-        return false
     }
 
     override fun lower(irFile: IrFile) {
         irFile.transformChildrenVoid(object : IrElementTransformerVoid() {
             override fun visitCall(expression: IrCall): IrExpression {
                 expression.transformChildrenVoid(this)
-                return if (isNegation(expression, context) && isNegation(expression.dispatchReceiver!!, context)) {
-                    (expression.dispatchReceiver as IrCall).dispatchReceiver!!
-                } else if (isNullCheckOfPrimitiveTypeValue(expression, context)) {
-                    val left = expression.getValueArgument(0)!!
-                    val nonNullArgument = if (left.isNullConst()) expression.getValueArgument(1)!! else left
-                    val constFalse = IrConstImpl.constFalse(expression.startOffset, expression.endOffset, context.irBuiltIns.booleanType)
-                    if (hasNoSideEffectsForNullCompare(nonNullArgument)) {
-                        constFalse
-                    } else {
-                        IrBlockImpl(expression.startOffset, expression.endOffset, expression.type, expression.origin).apply {
-                            statements.add(nonNullArgument.coerceToUnitIfNeeded(nonNullArgument.type.toKotlinType(), context.irBuiltIns))
-                            statements.add(constFalse)
+
+                if (isNegation(expression, context) && isNegation(expression.dispatchReceiver!!, context)) {
+                    return (expression.dispatchReceiver as IrCall).dispatchReceiver!!
+                }
+
+                getOperandsIfCallToEqeqOrEquals(expression)?.let {
+                    val left = it.first
+                    val right = it.second
+
+                    return when {
+                        left.isNullConst() && right.isNullConst() ->
+                            IrConstImpl.constTrue(expression.startOffset, expression.endOffset, context.irBuiltIns.booleanType)
+
+                        it.second.isNullConst() && it.first.type.unboxInlineClass().isPrimitiveType() ||
+                                it.first.isNullConst() && it.second.type.unboxInlineClass().isPrimitiveType() -> {
+                            val nonNullArgument = if (left.isNullConst()) right else left
+                            val constFalse =
+                                IrConstImpl.constFalse(expression.startOffset, expression.endOffset, context.irBuiltIns.booleanType)
+                            if (hasNoSideEffectsForNullCompare(nonNullArgument)) {
+                                constFalse
```

Comment:
It's for null == primitive, but null == "non-primitive" is lost cause of 'case' condition

Reply:
Bytecode test is required

Reply:
You are absolutely correct. Added bytecode tests to showcase presence of optimization (and the absence of a possible optimization when we find an easier way to pick out the default equals method.